fix: ODPS data source table preview failed #32301

zhutong6688 opened the pull request:
### SUMMARY
I fixed an error when using sqllab to preview data after configuring the odps data source in superset, and added partition restrictions for previewing SQL

fixes: https://github.com/apache/superset/issues/32301

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->
This is a preview of the odps table data before the repair：
![image](https://github.com/user-attachments/assets/7c5d3791-ee29-4b96-9597-5378a49b822f)

This is a preview of the repaired ODPS table data：
![5f49bcaba6d6cd31bdfdfae4ada6970](https://github.com/user-attachments/assets/a57f5188-63a5-45d3-92e8-6273de04f47e)
![d2de24aca26b1c8560deaabbf3d1144](https://github.com/user-attachments/assets/5fa61af1-a6c9-431d-a4ca-92f91b21ad0f)
![8aeca80279b56aaf4b3ee4ffeec96d5](https://github.com/user-attachments/assets/f76360ac-439a-4ede-a1df-6c916b76709d)
In addition, when using other data sources, there is no problem with testing and it does not affect other data sources

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->
1. First, configure the ODPS data source
2. Create a new SQLlab page in the "SQL" column, select the database Schema
3. Select partitioned tables, non partitioned tables, and multi partitioned tables in sequence to test the correctness of data preview
4. Choose another data source, such as an example data source like SQLite, and follow the same steps as above to verify the validity of the data preview

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

hainenber commented:
Thank you for the PR! I left a few comments and also you'll need to run added changes against established pre-commit hooks [here](https://superset.apache.org/docs/contributing/development/#git-hooks).

hainenber commented on superset/sql/parse.py:
```diff
@@ -124,6 +124,21 @@ def __str__(self) -> str:
+    partition_column: list | None = None
+    def __str__(self) -> str:
+        """
+        Return the partition columns of table name.
```
It'd be great if you have an example for stringified Partition object here

zhutong6688 replied:
You're right, I'll add an example later

hainenber commented on superset/sql/parse.py:
```diff
@@ -124,6 +124,21 @@ def __str__(self) -> str:
+    partition_column: list | None = None
+    def __str__(self) -> str:
+        """
+        Return the partition columns of table name.
```
It'd be great if you have an example for stringified Partition object here

zhutong6688 replied:
You're right, I'll add an example later

zhutong6688 commented on superset/daos/database.py:
```diff
@@ -165,7 +168,30 @@ def get_ssh_tunnel(cls, database_id: int) -> SSHTunnel | None:
+        uri = database.sqlalchemy_uri
+        passwd = database.password
+        pattern = re.compile(
+            r'odps://(?P<username>[^:]+):(?P<password>[^@]+)@(?P<project>[^/]+)/(?:\?endpoint=(?P<endpoint>[^&]+))'
```
Take a look at the back, this one is in use

hainenber replied:
I mean there is no `match.group('password')` to indicate usage. Is the naming for the captured group only for readability?

zhutong6688 commented on superset/daos/database.py:
```diff
@@ -165,7 +168,30 @@ def get_ssh_tunnel(cls, database_id: int) -> SSHTunnel | None:
-
+    
+    @classmethod
+    def is_odps_partitioned_table(cls, database: Database, table_name: str):
```
Are you saying adding some function descriptions or comments?

hainenber replied:
it's the return value's type. Would be something similar to below

```
def is_odps_partitioned_table(cls, database: Database, table_name: str) -> Tuple(bool, [str]]:
```

zhutong6688 commented on superset/databases/api.py:
```diff
@@ -1002,15 +1003,16 @@ def table_metadata(self, pk: int) -> FlaskResponse:
-
-        table = Table(parameters["name"], parameters["schema"], parameters["catalog"])
+        table_name = str(parameters["name"])
+        ispartitioned_table, partition_fields = DatabaseDAO.is_odps_partitioned_table(database,table_name)
```
Ispartitioned is for the convenience of determining whether the current table is partitioned, and also for the convenience of further expanding new functions in the future

hainenber replied:
I mean, `ispartitioned` doesn't follow usual Python snake-case naming convention. It usually goes like `is_partitioned_table`, fyi

zhutong6688 commented on superset/db_engine_specs/base.py:
```diff
@@ -1689,7 +1690,11 @@ def select_star(  # pylint: disable=too-many-arguments
         qry = select(fields).select_from(text(full_table_name))
-
+        if database.backend == 'odps':
+            if partition.ispartitioned_table and partition.partition_column is not None and len(partition.partition_column) > 0:
```
I will add a non null judgment later

zhutong6688 commented on superset/daos/database.py:
```diff
@@ -165,7 +168,30 @@ def get_ssh_tunnel(cls, database_id: int) -> SSHTunnel | None:
+            access_id = match.group('username')
+            project = match.group('project')
+            endpoint = match.group('endpoint')
+        access_key = passwd
```
Access_key=passwd, I assigned passwd to Access_key, so that when using ODPS to build connections later, I have a visual experience of AK and Secret

hainenber replied:
Are there any benefits aside of readability? I'm not holding strong opinion on this but feels like this assignment is a bit excessive.

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

rusackas commented:
approving a CI run, and adding a couple more reviewers for good measure :)

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 commented:
Please approve and agree a CI run，I have made the necessary corrections based on the error message

zhutong6688 commented:
Some of the data source test classes did not pass. Let me take a look at the problem again and see if there are any issues with the test classes being written

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
Merge branch 'apache:master' into fix-odps-partition-bug-32301

zhutong6688 commented:
Please approve and agree a CI run again, Thanks a lot

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 commented:
Could you please help me run a CI again so that I can see if there are any other check errors，Thanks again

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
Merge branch 'apache:master' into fix-odps-partition-bug-32301

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 commented:
please help me run a CI again，thanks a lot

zhutong6688 committed:
Merge branch 'apache:master' into fix-odps-partition-bug-32301

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 commented:
please help me run a CI again，thanks a lot

zhutong6688 commented:
![image](https://github.com/user-attachments/assets/efe092d7-fdfc-4fb7-90a9-b4f32f0fe978)
@hainenber May I ask if this error is caused by not installing and using pre commit?

hainenber commented:
I'd recommend to have precommit installed and used to catch early CI issues.

zhutong6688 commented:
> I'd recommend to have precommit installed and used to catch early CI issues.

Okay, got it

hainenber commented on requirements/base.in:
```diff
@@ -23,3 +23,4 @@ numexpr>=2.9.0
 # 5.0.0 has a sensitive deprecation used in other libs
 # -> https://github.com/aio-libs/async-timeout/blob/master/CHANGES.rst#500-2024-10-31
 async_timeout>=4.0.0,<5.0.0
+pyodps>=0.12.2
```
I don't see this library getting used in the PR. Can you provide some contexts?

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 commented:
@hainenber please help me run a CI again，thanks a lot

zhutong6688 commented:
Could you please help me agree and check the code online? Thank you

hainenber commented:
CI has already run. I’ll be reviewing your code.

zhutong6688 commented:
I found that the code for table preview has been greatly adjusted, which is exactly where I fixed it. The interface calls are different from what I modified before, and the only thing fixed is the improvement. It is suggested to slow down the improvement of this section of code, as there are indeed many bugs and bug fixes cannot keep up with the speed of improving the code. This is not reasonable, and it is more appropriate to prioritize bug fixes in the code.

zhutong6688 committed:
Merge branch 'apache:master' into fix-odps-partition-bug-32301

zhutong6688 committed:
Merge remote-tracking branch 'origin/fix-odps-partition-bug-32301' into fix-odps-partition-bug-32301

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 committed:
fix bug of ODPS data source table preview failed

zhutong6688 commented:
I have adjusted the code. Thank you for agreeing to run the constantly blocked reviews

zhutong6688 committed:
Merge branch 'apache:master' into fix-odps-partition-bug-32301

zhutong6688 commented:
@hainenber @rusackas @sadpandajoe @bgreenlee Let's take a look at this PR (# 32556) later. The previous PR (# 32310) modified a lot of underlying basic classes. Although other data sources were not affected according to my testing, the testing unit classes of superset need to be adjusted too much. Therefore, I submitted a new PR, which tries not to modify the underlying basic classes as much as possible, but modifies the logical collection to the odps.by file, without causing too much impact on a large number of unit testing classes.

zhutong6688 commented:
Let's pay attention to this PR（https://github.com/apache/superset/pull/32556） in the future. This PR will be closed for now. It has modified many underlying classes, which requires too many modifications to the test classes of the superset project itself。
